### PR TITLE
fix(expr): do not const-eval impure expressions

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously we treat an expression tree that only consists of `InputRef`s and function calls as constant evaluable. This is incorrect if there are impure function calls, for example, `PROCTIME()`. (#9572)

This PR also makes `eval_row_const` return an `Option` to avoid calling `is_const` multiple times.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/expr/mod.rs [hunk 1]
```diff
@@ -539,26 +548,41 @@ impl ExprImpl {
...
-        }
-        impl ExprVisitor<()> for Has {
-            fn merge(_: (), _: ()) {}
-
-            fn visit_expr(&mut self, expr: &ExprImpl) {
-                match expr {
-                    ExprImpl::Literal(_inner) => {}
-                    ExprImpl::FunctionCall(inner) => self.visit_function_call(inner),
-                    _ => self.has = true,
+        let can_eval_const = {
```
Comment (st1page) [inline src/frontend/src/expr/mod.rs:554]:
```suggestion
        let input_related = {
```

File: src/frontend/src/expr/mod.rs [hunk 2]
```diff
@@ -276,12 +276,21 @@ impl ExprImpl {
...
-        self.eval_row(&OwnedRow::empty())
-            .now_or_never()
-            .expect("constant expression should not be async")
+    /// Try to evaluate an expression if it's a constant expression by `ExprImpl::is_const`.
+    ///
+    /// Returns...
+    /// - `None` if it's not a constant expression,
+    /// - `Some(Ok(_))` if constant evaluation succeeds,
+    /// - `Some(Err(_))` if there's an error while evaluating a constant expression.
+    pub fn eval_row_const(&self) -> Option<RwResult<Datum>> {
```
Comment (liurenjie1024) [inline src/frontend/src/expr/mod.rs:285]: To follow rust's api name convention, we should rename this to `try_eval_row_const`?

File: proto/expr.proto [hunk 3]
```diff
@@ -168,9 +168,6 @@ message ExprNode {
     JSONB_TYPEOF = 602;
     JSONB_ARRAY_LENGTH = 603;
 
-    // Functions that return a constant value
-    PI = 610;
```
Comment (liurenjie1024) [inline proto/expr.proto:172]: Why remove this?

File: src/frontend/src/expr/mod.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/expr/mod.rs:285]: how about `try_const_folding`?

Comment (st1page) [reply src/frontend/src/expr/mod.rs:285]: It is not very related to this PR. I will do it later.

File: proto/expr.proto (hunk 3 above)
Comment (BugenZhao) [reply proto/expr.proto:172]: This is dead code. 🥵

File: src/frontend/src/expr/mod.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/expr/mod.rs:285]: `try_fold_constant`

Comment (BugenZhao) [reply src/frontend/src/expr/mod.rs]: Done

**Review by liurenjie1024 — APPROVED** [review]
LGTM

**Pull request merged**